server/index.js

```diff
@@ -95,9 +90,17 @@ export default class Server {
     }
   }
 
-  async render (req, res, pathname, query) {
-    const html = await this.renderToHTML(req, res, pathname, query)
-    sendHTML(res, html)
+  async render (req, res, pathname, query, use) {
+    const accept = accepts(req)
+    if (accept.type(['json', 'html']) === 'json') {
+      // When we are asking for the page json, we always need to give
+      // the real pathname. But not the pathname user provided.
+      const { pathname: realPathname } = parse(req.url)
```

Comment:
Now we've to parse the url twice to get the pathname. But we don't need query params just like in the other case.

This might add some overhead but it's negligible because of following reasons:

* Now we don't parse accept headers twice
* Rendering the React component on the server is so much costly than this.